File: log-view-machine/java/src/main/java/com/logview/causality/StateMachine.java
package com.logview.causality;

/**
 * Immutable state machine definition (XState-congruent).
 * Created via StateMachine.create(config).
 */
public final class StateMachine<T> {

    private final MachineConfig<T> config;

    private StateMachine(MachineConfig<T> config) {
        this.config = config;
    }

    public static <T> StateMachine<T> create(MachineConfig<T> config) {
        return new StateMachine<>(config);
    }

    public String getId() { return config.getId(); }
    public String getInitial() { return config.getInitial(); }
    public T getContext() { return config.getInitialContext(); }
    public MachineConfig<T> getConfig() { return config; }
}

File: log-view-machine/core/src/main/kotlin/com/logview/causality/StateMachine.kt
package com.logview.causality

/**
 * Immutable state machine definition (XState-congruent).
 * Created via createMachine(config).
 */
class StateMachine<T> private constructor(
    val config: MachineConfig<T>
) {
    val id: String? get() = config.id
    val initial: String get() = config.initial
    val context: T get() = config.getInitialContext()

    companion object {
        fun <T> create(config: MachineConfig<T>): StateMachine<T> =
            StateMachine(config)
    }
}

/** Convenience: createMachine(config) */
fun <T> createMachine(config: MachineConfig<T>): StateMachine<T> =
    StateMachine.create(config)
